### ISSUE TITLE
```EarlyTask``` plugins

### ISSUE THREAD

**Issue by felixfontein** [getnikola/nikola#1562]
Hi,

I'd like to have a plugin category `EarlyTask`, for tasks which are executed before the site is rendered (i.e. an analogue to `LateTask`). I personally need that for a plugin (or better, combination of plugins) I wrote, currently I used the `Task` plugin class but it happens that some tasks are run after page compiling, while my page compiling plugin needs their result -- and so it fails.

Does anyone mind if I add something like that? Or would it be better to have a general priority system, so you can assign a task a priority (usual tasks could get 10, and late tasks 100, so you could add a task with priority 2 and one with priority 7 to ensure that the one with priority 2 appears in the task list before the one with priority 7 and before all regular rendering tasks and late tasks)?

(I have the vague feeling that I already read something about `EarlyTask`s somewhere here, but I cannot remember where. So it's probably not my own idea :) )

Cheers,
Felix

**Comment by felixfontein** [comment]
Cool, everything's already done! :)
A priority system won't be exactly complicated, since most things can have the same priority -- there are still doit's dependencies to handle that. But then, that was just a suggestion, I'm perfectly happy with ```EarlyTask``` as it is now. Thanks!
Well, and of course, a Happy New Year to you all!

**Comment by schettino72** [comment]
Probably too late but i will chime in anyway :)

doit has 2 distinct phases. task-creation and task-execution.

task-creation (generate task metadata) is usually done in doit using the task_xxx functions ina dodo.py module. But in Nikola it is done through the Nikola site object. Since in Nikola you can add more task-creators through plugins `LateTask` was created to make sure the Nikola site object was set-up before executing task-creators from plugins... 

So `LateTask` means late-generated-task not late-executed-task. `LateTask` is nikola concept and `doit` has no idea about it.

task-execution ordering must be handled by doit using the task property `task_dep`. 

A "priority system" may make sense for Nikola since the global Nikola site object is modified by many plugins, but it doesnt make sense for `doit` itself.

**Comment by felixfontein** [comment]
What do you think? Would it be hard to modify ```DoitNikola``` (which inherits from ```DoitMain```) and ```NikolaTaskLoader``` (inherited from ```TaskLoader```) to get the following behavior:

 * The set of tasks are split up into three (or more) categories: early, usual, late.
 * For all commands (except ```build```), the task loader will behave as it is now, i.e. generate tasks for early, then for usual, then for late, and return a big list containing all tasks.
 * The ```build``` command behaves differently, by first creating only the early tasks and running them all, *then* creating the usual tasks, running them all, and finally creating the late tasks, and running them all.
    * For this, it might be possible to call the inherited ```run``` method of ```DoitMain``` three times, each time with a different task loader (one only giving the early tasks, one only giving the usual tasks, etc. -- it could also be the same task loader with a enum ```ALL```, ```EARLY```, ```USUAL```, ```LATE```)
    * Maybe it would also be better to create three ```DoitMain``` objects, one for each set of tasks.

What do you think?

Cheers,
Felix

**Comment by ralsina** [comment]
@felixfontein it's perhaps cleaner and less work to create three "metatasks" and have mt1 have all EarlyTasks in its task_dep, then have mt2 have all Tasks and mt1 on its task_dep and then have mt3 have all LateTasks and mt2 in its task_dep

It can be trivially extended to an enumeration etc.

**Comment by Kwpolska** [comment]
I’m reverting the current failed implementation of EarlyTask in 0ce8d72.

**Comment by felixfontein** [comment]
@ralsina: That does not solve the problem that during creation of the usual tasks, the generated posts haven't been created yet (and so no tasks to handle them, i.e. to render their pages, include them in indices and tag pages, etc. can be created resp. are created incorrectly). For that, the tasks can only be created when the previous tasks have been completely processed.

**Comment by ralsina** [comment]
@felixfontein if a task needs another task to be executed to get the information it needs, then it should get that information on runtime, not on creation time. It can even obtain that information from the earlier task using something like http://pydoit.org/dependencies.html#calculated-dependencies

**Comment by ralsina** [comment]
Or maybe something like http://pydoit.org/dependencies.html#getargs

**Comment by felixfontein** [comment]
The problem is that at creation time of the page compilation tasks, they have no idea if (or which) new posts are created. But the tasks (just to compile the post to .html, for example) must be created at that point, for the specific posts which will be created (and whose names are not known to Nikola at that point).

**Comment by ralsina** [comment]
Or, maybe, what should happen is that the post scanning should be a task, which it's not.

**Comment by felixfontein** [comment]
Even if post scanning is a task, you have a problem since after running that task, new tasks have to be created.

**Comment by ralsina** [comment]
Yes, there is a large semantic hole here. What we need is to figure out a way (which probably involves some largish changes in Nikola) so that tasks can affect "the future".

IOW: 

1) a task modifies the timeline by creating a post
2) that means post scanning has to happen
3) that means more tasks have to be created for the new timeline

If we can work out how to do that in a clean manner, the rest will just work. Of course this is not trivial or it would be there already :-)

**Comment by felixfontein** [comment]
An ideal workflow would look like:
 * create tasks for ```EarlyTask``` plugins,
 * run tasks for ```EarlyTask``` plugins,
 * scan posts,
 * create tasks for ```Task``` plugins,
 * create tasks for ```LateTask``` plugins,
 * run tasks for ```Task``` and ```LateTask``` plugins.

(The last three steps can also be split into two runs: first create and execute ```Task``` tasks, then create and execute ```LateTask``` tasks.)

**Comment by felixfontein** [comment]
Well, by having more than one "create and then execute tasks" steps, this can be done in a quite simple way.

**Comment by ralsina** [comment]
Until we need to have a task that runs between EarlyIshTasks and NotSoEarlyTasks

**Comment by ralsina** [comment]
Ok, I give up, let's do that separation and multiple runs.

**Comment by felixfontein** [comment]
Well, that's why I proposed having a priority system, where priority indicates in which batch the tasks are created and executes. The current system (including ```EarlyTask```s) would have three priorities.

**Comment by schettino72** [comment]
I think Nikola running doit three (or more) times will be too messy, and will create many other problems. I suggest to do some work on doit before trying to implement this on nikola.

First lets try to come up with a simple example that does not use nikola code base but has a similar workflow. Then change doit to handle that, and finally use it on nikola...

**Comment by felixfontein** [comment]
Can you give me some hints what might go wrong? Are there certain things in doit that should not be used/initialized more than once?

**Comment by felixfontein** [comment]
Ok, I created a small test program: https://github.com/getnikola/nikola/blob/earlytask_experiments/test.py

It creates (and destroys pre-existing) files '1' and '2' and a directory 'dest', so be a bit cautious when running it.

On the first glance, seems to work well.

**Comment by schettino72** [comment]
> Can you give me some hints what might go wrong?

 - try executing only one task. you will need to some pre analysis of available tasks or error handling.
 - task_dep: you will need to make sure to only use tasks on the same level (on different level they will be managed by nikola itself)
 - clean command: also will need some changes because it is executed in reverse order.
 - auto command (original one from doit): i guess would need a lot of work
 - output: as of today doit and nikola output is very simple so no problem. but it will be an issue if there is a summary of execution at the end (It is on my TODO list)
 - task parameters `getargs`, `setup`, `teardown` will all have issues (it seems nikola doesnt use any of this but not nice to make it permanently unavailable.)
 - doit at the moment has very poor tooling but that may change in the future, this approach will probably make nikola not be able to use those.

> Ok, I created a small test program: https://github.com/getnikola/nikola/blob/earlytask_experiments/test.py

It is cool you could do that given doit limitations... I also want to give it a try changing doit, but it wont be as fast as you :) do you want to help me?

**Comment by schettino72** [comment]
@felixfontein so I created a proof of concept.  (https://github.com/pydoit/doit/compare/delayed-task-creation)

I created an issue on doit tracker for this: pydoit/doit#14 .
It is still WIP but can you take a look at it and check if this would solve your problem with nikola.

**Comment by schettino72** [comment]
> The approach you want to chose is a bit more compliated, but should be doable. I'll take a closer look at your branch during the next days...

Maybe my approach is more complicated to implement but it is the Right Thing :tm:
And the most complicated part was already done long time ago when doit added support to `calc_dep`... 

As of the code base now I am sure it will be easier to implement on doit than on nikola. I also expected to add this feature soon or later, so regardless nikola uses it or not it will be added to doit.

It also brings the possibility of some extra benefits as this is the foundation to avoid task-creation of all tasks when they are not required. I.e. nikola implemented several stuff (serve, check, deploy...) as commands to avoid the costly scan-posts.

**Comment by felixfontein** [comment]
Yes, it's the Right Thing :-)

Nikola's ```check``` does need to scan the posts to know which files to inspect or which files shouldn't be there (because they're not created by a task) -- even though it does so only by calling ```nikola list --all``` internally.

**Comment by schettino72** [comment]
doit part to handle this is ready :) already merged to master... https://github.com/pydoit/doit/commit/b06eb8ebe0d5265a6221900f80905aaf0868d217

The change above includes docs but since nikola doesnt use dodo.py it is required to use
some undocumented internal API... 

Where is the plugin that requires this change? I guess the best way to integrate it on nikola is:
 
1) create a task with no actions like "pre_build"
2) the creation of build tasks should be delayed to be done after "pre_build" is executed
3) plugins will need to inject themselves as a task_dep of "pre_build".

**Comment by felixfontein** [comment]
Cool!

As a first experiment, I'd suggest to use this to actually generate ```LateTask``` tasks after ```Task``` tasks have been executed. When adding ```EarlyTask```s (or whatever else), it can be added exactly the same way.

The main part happens in ```nikola/__main__.py```, in the class ```NikolaTaskLoader```:
```
class NikolaTaskLoader(TaskLoader):
    """custom task loader to get tasks from Nikola instead of dodo.py file"""
    def __init__(self, nikola, quiet=False):
        self.nikola = nikola
        self.quiet = quiet

    def load_tasks(self, cmd, opt_values, pos_args):
        if self.quiet:
            DOIT_CONFIG = {
                'verbosity': 0,
                'reporter': 'zero',
            }
        else:
            DOIT_CONFIG = {
                'reporter': ExecutedOnlyReporter,
                'outfile': sys.stderr,
            }
        DOIT_CONFIG['default_tasks'] = ['render_site', 'post_render']
        tasks = generate_tasks(
            'render_site',
            self.nikola.gen_tasks('render_site', "Task", 'Group of tasks to render the site.'))
        latetasks = generate_tasks(
            'post_render',
            self.nikola.gen_tasks('post_render', "LateTask", 'Group of tasks to be executed after site is rendered.'))
        signal('initialized').send(self.nikola)
        return tasks + latetasks, DOIT_CONFIG
```
So I guess that we have to do the following:
 * add all tasks in ```tasks``` to be dependencies of a new task, ```late_build```, who does nothing;
 * when ```late_build``` is executed (i.e. everything else is done), ```latetasks``` is generated and its tasks are executed.

So something along:
```
        ...
        tasks = generate_tasks(
            'render_site',
            self.nikola.gen_tasks('render_site', "Task", 'Group of tasks to render the site.'))
        tasks.extend(
            generate_tasks('late_build',
                { 'basename': 'late_build',
                  'task_deps': tasks })
        tasks.extend(
            generate_tasks('late_build',
                { 'basename': 'late_build_create',
                  'create_after': 'late_build',
                  ...
#        latetasks = generate_tasks(
#            'post_render',
#            self.nikola.gen_tasks('post_render', "LateTask", 'Group of tasks to be executed after site is rendered.'))
                  ... })
        signal('initialized').send(self.nikola)
        return tasks, DOIT_CONFIG
```

**Comment by ralsina** [comment]
Just a heads up for whoever tries: This *may* break "nikola check"

**Comment by felixfontein** [comment]
Actually, ```nikola check``` should not be affected, since for all commands except ```build```/```run```, simply all tasks will be created. Or at least that's what I remember from schettino72's code.

**Comment by ralsina** [comment]
If that's so, even better, just a thing to check :-)

**Comment by felixfontein** [comment]
I started implementing here: https://github.com/getnikola/nikola/tree/earlytask_impl
So far I'm mostly preparing stuff, the aim is to use the new doit features similarly as in https://github.com/getnikola/nikola/blob/earlytask_experiments/test.py.

In the first step, I also moved post scanning into an own plugin. I've also changed what I called 'level' in that test to 'stage'; I hope that's less confusing.

**Comment by felixfontein** [comment]
The basic stuff works, I think: ```build```, ```clean```, ```list``` (and thus ```check```), ...

What does not work yet: if you specify a specific task/target with ```build``` which isn't in the first stage created. Also, it is not clear when the ```initialized``` signal should be emitted, since there is not any longer a point when all tasks are created but none is executed.

**Comment by felixfontein** [comment]
For ```initialized```: what about having a signal for every stage? I.e. when the tasks for stage _n_ have been created, emit the signal ```initialized_n``` (with nicer names for special _n_, like -10, 1, 10, 100, corresponding to the phases _early_, _post scanning_, _site rendering_, and _late_). The classic ```initialized``` signal could be emitted for _n_ = 10, i.e. after generation of the usual site rendering tasks.

Is anyone actually using that signal at the moment?

**Comment by felixfontein** [comment]
For the other problem (specifying tasks/targets on command line): I don't see a good solution here which is not somewhat hacky.

The simplest solution would be to create tasks for all stages until all targets and tasks mentioned in the command line are created, and then let doit decide which tasks to execute. (default_tasks isn't of relevance anymore at this time.)

Has anyone an opinion on this? @schettino72 maybe?

**Comment by felixfontein** [comment]
Also, another thing: running in parallel doesn't work anymore out of the box, since multiprocessing doesn't work well with delayed task creation. But if you use threads instead of processes (by specifying ```-P thread```), it will run fine in parallel.

**Comment by schettino72** [comment]
link to felix's branch: https://github.com/getnikola/nikola/compare/earlytask_impl

Felix Can you squash your commits? It makes easier to write comments. Now I need to find which commit did what change and the comments would be spread out...

> For the other problem (specifying tasks/targets on command line): I don't see a good solution here which is not somewhat hacky.

I guess the problem is that `Nikola.gen_tasks()` generate all tasks from a given plugin stage. It should be modified in a way that you have a function to create tasks from a given plugin instance (not from a plugin stage). So on `NikolaTaskLoader._gen_stage()` you would create a task for every plugin instance and not only one task for each stage. 

I am not 100% convinced about this usage of "stages"... why not just directly depend on task/plugin name?

> Also, another thing: running in parallel doesn't work anymore out of the box, since multiprocessing doesn't work well with delayed task creation.

The problem is usage of closures (or whatever that can not be pickled) with multiprocessing. There is no real reason to use closures (apart for small convenience of the code location). Someone could just refactor nikola code to avoid closures and multiprocessing would work fine with DelayedTasks.

**Comment by felixfontein** [comment]
It would be nice to create tasks for individual task names, but the problem is that you don't know which basenames a plugin will emit before asking it to create its tasks. (The galleries plugin usually outputs ```render_galleries``` tasks, but also ```_render_galleries_clean``` tasks; the sitemap plugin outputs ```sitemap``` (excepted) and ```_scan_locs``` tasks, and the gzip plugin outputs names depending on which task it processes -- that's somewhat different though, since its a ```TaskMultiplier``` plugin. And then there might be 3rd party plugins which might rely on that, too.)

@Kwpolska, @ralsina, what do you think? Would it be ok to make the assumption that that ```plugin.name``` is the only task name callable via ```nikola build <taskname>```?

A second thing: what happens if you use ```doit run bla```, where ```bla``` is a target of a task which is generated on the fly? It doesn't work with doit (I tested with your example: https://gist.github.com/schettino72/9868c27526a6c5ea554c), and so won't work with Nikola either. This could only be fixed by some kind of hack

**Comment by felixfontein** [comment]
Regarding avoiding pickling/enabling multiprocessing: the main culprit here is ```config_changed```, I think. Avoiding that will be hard to impossible without rewriting _a lot_ of code, including 3rd party code, I think.

**Comment by felixfontein** [comment]
(I now squashed the commits.)

**Comment by Kwpolska** [comment]
> Would it be ok to make the assumption that that plugin.name is the only task name callable via `nikola build <taskname>`?

`nikola build foo` generates and executes (if necessary) `foo` tasks and **all their dependencies**.  In other words, `nikola build sitemap` ≈ `nikola build` in standard sites.

You can also call *any* task basename from `nikola build`.  Including those which do not have the same name as their providing plugin.

**Comment by ralsina** [comment]
I adhere to schettino's question:

"I am not 100% convinced about this usage of "stages"... why not just directly depend on task/plugin name?"

We could make all plugins emit one task with the plugin name depending on all its previous tasks, if that helps.

**Comment by schettino72** [comment]
> A second thing: what happens if you use doit run bla, where bla is a target of a task which is generated on the fly?

You are right specifying the `target` of a DelayedTask wont work. 
This is a hard to solve problem. I see a few options but I am not sure they are worth the trouble:

  * when the specified argument is not a known task or target, doit would load delayed tasks (as done by the list command) to try to find the target. This would not be 100% reliable and might bring other problems...
  * query the DB looking for the `task` that generated the `target`, this would require loading and searching the whole DB because as of now the DB have efficient lookup only by task name.
  *  have some kind of rule/regex that would be able to find a task for given target path.

**Comment by felixfontein** [comment]
@ralsina: but what if you want to build only a task which is not equal to the plugin's name? And not everything belonging to that plugin? That won't be possible with approach.

**Comment by felixfontein** [comment]
@schettino72: Actually, I like the first idea a lot. I mean, it won't be 'worse' than it is now (concerning the order of execution and task generation).

We could also restrict the syntax of ```nikola build <...>``` to only allow to specify plugin names or something of the kind ```plugin_name:target```.

**Comment by schettino72** [comment]
> @schettino72: Actually, I like the first idea a lot. I mean, it won't be 'worse' than it is now (concerning the order of execution and task generation).

@felixfontein  created https://github.com/pydoit/doit/issues/20 with some further thoughts about it. Hopefully you can assign yourself to implement it :)

**Comment by felixfontein** [comment]
I'll try :) Though not today anymore...

**Comment by felixfontein** [comment]
Ok, I now rewrote parts of the code to have each task plugin's tasks generated by one delayed task loader. Also, the delayed task's name equals the task plugin's name, whence ```nikola build <task_name>``` works again.

One thing I noticed: since all tasks of stage 2 (f.e.) depend on the waiting task of stage 1, and that waiting task depends on all tasks of stage 1, building one specific task in stage 2 via ```nikola build <task_name>``` triggers a build of all tasks of stage 1. This could be helped by adding a modified version of ```task_dep``` to doit, which is only used to determine the order of execution, but not which tasks have to also be built before a specified task can be build. @schettino72: what do you think about this?

**Comment by schettino72** [comment]
> This could be helped by adding a modified version of task_dep to doit, which is only used to determine the order of execution, but not which tasks have to also be built before a specified task can be build. @schettino72: what do you think about this?

Do you mean a [setup-task](http://pydoit.org/dependencies.html#setup-task)?
uhmm. The docs need an example without a `teardown`.

Maybe a delayed task should create an implicit setup-task instead of a task_dep...
It is a trivial change, can you try it?

**Comment by felixfontein** [comment]
No, a setup-task will be executed when this task is executed. A ```wait-for``` dependency should _not_ be executed (except of course it is manually specified on the command line, or it also appears as a proper dependency of another task to be executed), it should only participate in determining the execution order resp. when to start executing a task.

**Comment by schettino72** [comment]
@felixfontein give me an example please. dodo.py format and what helps when you run it. better create an issue on doit tracker or we gonna hijack this issue (again).

**Comment by felixfontein** [comment]
Take the following ```dodo.py``` file:
```
def task_a_start():
    return {
        'basename': 'a_start',
        'actions': None,
    }

def task_a1():
    return {
        'basename': 'a1',
        'task_dep': ['a_start'],
        'actions': ['echo A1'],
    }

def task_a2():
    return {
        'basename': 'a2',
        'task_dep': ['a_start'],
        'actions': ['echo A2'],
    }

def task_a_wait():
    return {
        'basename': 'a_wait',
        'task_dep': ['a1', 'a2'],
        'actions': None,
    }

def task_b_start():
    return {
        'basename': 'b_start',
        'task_dep': ['a_wait'],
        'actions': None,
    }

def task_b1():
    return {
        'basename': 'b1',
        'task_dep': ['b_start'],
        'actions': ['echo B1'],
    }

def task_b2():
    return {
        'basename': 'b2',
        'task_dep': ['b_start', 'a2'],
        'actions': ['echo B2'],
    }

def task_b_wait():
    return {
        'basename': 'b_wait',
        'task_dep': ['b1', 'b2'],
        'actions': None,
    }
```
There are two stages, ```a``` and ```b```. To ensure that ```b``` is executed when ```a``` is done, ```a_wait``` depends on all ```a``` tasks, all ```b``` tasks depend on ```b_start```, and ```b_start``` depends on ```a_wait```. There's also a dependence between ```b2``` and ```a2```.

I would like this last dependence (of ```b_start``` on ```a_wait```) to be a ```wait-for``` dependence, so that if I run ```doit b1```, only ```a1``` (and ```a_start```) are executed. (And if I run ```doit b2```, only ```a2``` and ```b2``` and the corresponding ```_start``` tasks are executed.)

**Comment by felixfontein** [comment]
I think for first discussing on how to do this (because it has to do a lot with this feature) it's ok to discuss it here, but as soon as we know what we want we can continue to discuss it in the doit tracker. Hope that's ok for you :)

**Comment by schettino72** [comment]
@felixfontein thanks for the example. I guess I understand your problem

In my opinion this problem only arises when using "phases" that doit has really no support for, so maybe a patch on Nikola is more appropriate.

Can you define better what triggers the change of behaviour in these `wait-for` dependency? Is it when *any* task is specified in the command line? Sounds too tricky to me...

And how can you test/trigger this before https://github.com/pydoit/doit/issues/20 being implemented?

Anyway I gave it a try here:
https://github.com/schettino72/nikola/compare/getnikola:earlytask_impl...earlytask?expand=1 
Luckily I added `pos_args` in the signature of `load_tasks` even that I didnt know any use for it up to now :)

**Comment by felixfontein** [comment]
Hmm, a ```wait-for``` instead of ```task_dep``` could also be of interest if you want to process tasks in parallel, but some tasks need a resource which cannot be used in parallel (maybe some external device, like a DVD writer). For such a setup, you need a mechanism to make a second task to be not executed until a first task is done, but you don't want an explicit dependency so you can build each one individually.

Yes, I know that this sounds a bit far fetched, but at least it shows such a feature could _in theory_ be used in a more general setting.

Anyway, there's no behavior difference for ```wait-for``` for special situations; it should always behave the same way: if two tasks ```a``` and ```b``` are scheduled to be executed, and ```b``` ```wait-for``` ```a```, then ```a``` is not executed before ```b``` is done. So if ```b``` is specified as a task to be executed (either via command line or as a default task), this does not trigger a check of ```a```'s dependencies (to determine whether it should be executed) like a ```task_dep``` does. It only ensures that if ```a``` is actually executed, ```b``` will only be executed when ```a``` is done.

**Comment by felixfontein** [comment]
(Your try is a hack which works fine if all tasks specified on the command line are within one stage, but if they are not, tasks from a later stage might be generated before an earlier stage finished execution.)

**Comment by felixfontein** [comment]
Ok, I got an idea where this could be quite useful. Assume that you want to record audio samples, maybe for a study. Every sample (recorded as a .wav file) should be converted to different formats (say .ogg and .mp4) afterwards. So you create a recording task for every sample to record, and tasks to create .ogg and .mp4 files (which depend on the recording task). Since the encoding can be done in parallel, you want to run doit with ```-n2```. But you cannot record two things at the same time, so you need to introduce dependencies between the sample recordings.

If you have three recordings, ```a```, ```b``` and ```c```, you could use ```task_dep``` to get a chain ```a -> b -> c```. But now, if you only want to do recording ```b``` (for example because you noticed the recording has too much background noise), you want to run ```doit run b -n2```. But since there's a task dependence, doit will by default also execute task ```a```. So you end up doing two recordings, even though you needed only one.

Here you would prefer to use a ```wait-for``` dependency between ```a```, ```b``` and ```c```, and not a ```task_dep```.

(Even if you don't want to do the encoding part and thus don't need parallel execution, having such a ```wait-for``` dependency makes sense to protect against failures when once executing the tasks in parallel -- which might happen if your dodo.py is included in a larger dodo.py which does a lot more.)

**Comment by schettino72** [comment]
@felixfontein the audio sample example is a different case... this is an example where you need contention based on resource utilization for parallel scheduling. This has been raised before, it is feasible to be implemented in doit. But using `wait-for` would be a poor solution because it can handle just 1 resource being shared.

> if two tasks a and b are scheduled to be executed, and b wait-for a, then a is not executed before b is done. 

I guess you need to understand a bit more how doit works internally. A task is schedule to be executed in 2 situations:

 1) the task is specified in the command line or default_tasks
 2) the task is a dependency of a previously scheduled task

The problem is that in 2) this happens at run time while tasks are being executed. In other words, doit does not pre-compute the whole task dependency tree before it starts its execution. This has some advantages: being fast (dont compute parts of the tree that are not used), and allowing some dynamic modification of the "tree" (like calc_dep and delayed task creation).

To implement A `wait-for` B doit would have to finish all other scheduled tasks to make sure that none of the scheduled tasks would have a real task_dep to B. But since you might have multiple uses of `wait-for` even that would not guarantee that no further scheduled task would not have a task_dep on B. Thats kind of the same problem as you pointed in my hacky patch.

The other option would be to pre-compute the whole DAG, but again given the very dynamic nature of doit you would still have no guarantee that a "skipped" `wait-for` task would not be scheduled later by a third task. 

So I guess this `wait-for` could only be implemented if there was no support for dynamic changes in the task dependency-tree. Or do you have an idea on how an implementation would work? 

I guess it does not solve your problem but doit has a "--single" flag for the `run` command (or `build` in Nikola) that ignores task_dep. Sometimes useful to avoid rebuilding a lot of stuff when trying some changes.

**Comment by felixfontein** [comment]
Having ```--single``` is probably enough for most use-cases in Nikola (the need to avoid earlier stages to check for tasks to make is somewhat special anyway, I think). So maybe we can just ignore this thing. If I get a good idea how this could be solved/implemented while working on #20 in the doit tracker I'll try that out; if not let's leave it as it is. (Or does anyone objects to this?)

**Comment by felixfontein** [comment]
(Actually, that was why I used `task_dep` in the first place. It resulted in way too many things built, but it didn't stop with this problem :) )

**Comment by Kwpolska** [comment]
That’s not going to happen any time soon.
